Code/GraphMol/RWMol.cpp

```diff
@@ -574,26 +576,274 @@ void RWMol::beginBatchEdit() {
+                // github #6900 if we remove stereo atoms we need to remove
+                //  the CIS and or TRANS since this requires stereo atoms
+                if (obnd->getStereo() == Bond::BondStereo::STEREOCIS ||
+                    obnd->getStereo() == Bond::BondStereo::STEREOTRANS) {
+                    obnd->setStereo(Bond::BondStereo::STEREONONE);
+                }
+                obnd->getStereoAtoms().clear();
+            }
+        }
+        
+        removeSubstanceGroupsReferencingBond(*this, idx);
+        
+        bnd->setOwningMol(nullptr);
+        
+        MolGraph::vertex_descriptor vd1 = boost::vertex(aid1, d_graph);
+        MolGraph::vertex_descriptor vd2 = boost::vertex(aid2, d_graph);
+        boost::remove_edge(vd1, vd2, d_graph);
+        delete bnd;
+        --numBonds;
+    }
+    
+    // loop over all bonds with higher indices than the minimum modified and update their indices
+    ROMol::EDGE_ITER firstB, lastB;
+    boost::tie(firstB, lastB) = this->getEdges();
+    unsigned int next_idx = min_idx;
+    while (firstB != lastB) {
+        Bond *bond = (*this)[*firstB];
+        if (bond->getIdx() > min_idx) {
+            bond->setIdx(next_idx++);
+        }
+        ++firstB;
+    }
+}
+
+void RWMol::batchRemoveAtoms() {
+    if(!dp_delAtoms)
+        return;
+    std::vector<Atom*> oldIndices;
+    oldIndices.reserve(getNumAtoms());
+    for(auto *atom: atoms()) {
+        oldIndices[atom->getIdx()] = atom;
+    }
+    
+    auto &delAtoms = *dp_delAtoms;
+    for (unsigned int i = rdcast<unsigned int>(delAtoms.size()); i > 0; --i) {
+        if( !delAtoms[i-1] )
+            continue;
+        unsigned int idx = i-1;
+        Atom * atom = getAtomWithIdx(idx);
+        if (!atom)
+            continue;
+        
+        // remove any bookmarks which point to this atom:
+        ATOM_BOOKMARK_MAP *marks = getAtomBookmarks();
+        auto markI = marks->begin();
+        while (markI != marks->end()) {
+            const ATOM_PTR_LIST &atoms = markI->second;
+            // we need to copy the iterator then increment it, because the
+            // deletion we're going to do in clearAtomBookmark will invalidate
+            // it.
+            auto tmpI = markI;
+            ++markI;
+            if (std::find(atoms.begin(), atoms.end(), atom) != atoms.end()) {
+                clearAtomBookmark(tmpI->first, atom);
+            }
+        }
+        
+        // now deal with bonds:
+        //   their end indices may need to be decremented and their
+        //   indices will need to be handled and if they have an
+        //   ENDPTS prop that includes idx, it will need updating.
+        EDGE_ITER beg, end;
+        boost::tie(beg, end) = getEdges();
+        std::string sprop;
+        while (beg != end) {
+            Bond *bond = d_graph[*beg++];
+            if (bond->getPropIfPresent(RDKit::common_properties::_MolFileBondEndPts,
```

Comment:
I'd love to make this a free or separate function:

```
updateMolFileBondEndPts(btw, idx);
```

Reply:
yeah, we need to find a better solution for this at some point